File: src/main/java/ru/mydesignstudio/database/metadata/extractor/extractors/reference/ReferenceExtractor.java
package ru.mydesignstudio.database.metadata.extractor.extractors.reference;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.extractors.ExtractHelper;
import ru.mydesignstudio.database.metadata.extractor.extractors.TableMetadataAppender;
import ru.mydesignstudio.database.metadata.extractor.extractors.model.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.resource.StringResource;

@Component
public class ReferenceExtractor implements TableMetadataAppender {
  @StringResource("classpath:sql/extract_references.sql")
  private String extractQuery;

  @Autowired
  private ExtractHelper helper;

  @Override
  public void append(TableMetadata metadata, String schemaName, String tableName) {
    metadata.setReferences(helper.extract(extractQuery, new Object[]{ tableName, schemaName }, ReferenceModel.class));
  }
}

File: extractor-source/source-oracle/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/table/extractor/ReferenceExtractor.kt
package ru.mydesignstudio.database.metadata.extractor.table.extractor

import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.ExtractorBase
import ru.mydesignstudio.database.metadata.extractor.connection.DataSourceProvider
import ru.mydesignstudio.database.metadata.extractor.extract.helper.QueryProvider
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataObject
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSchema
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSourceConnection
import ru.mydesignstudio.database.metadata.extractor.extract.result.ReferenceModel

@Component
class ReferenceExtractor (
        private val dataSourceProvider: DataSourceProvider,
        private val queryProvider: QueryProvider) : ExtractorBase() {
    fun extract(metadata: MetadataObject, schema: MetadataSchema, connection: MetadataSourceConnection): Collection<ReferenceModel> {
        val query = queryProvider.provide("classpath:sql/extract_references.sql")
        val dataSource = dataSourceProvider.provide(connection)
        return extract(dataSource, query, listOf(metadata.name, schema.name)) { rs ->
            ReferenceModel(
                    childTable = getString(rs, "child_table"),
                    childColumn = getString(rs, "child_column"),
                    constraintName = getString(rs, "constraint_name"),
                    parentTable = getString(rs, "parent_table"),
                    parentColumn = getString(rs, "parent_column")
            )
        }
    }
}